firedrake/slate/preconditioners.py

```diff
@@ -0,0 +1,285 @@
+"""This module provides custom python preconditioners utilizing
+the Slate language.
+"""
+
+from __future__ import absolute_import, print_function, division
+
+import ufl
+
+from firedrake.matrix_free.preconditioners import PCBase
+from firedrake.petsc import PETSc
+from firedrake.slate.slate import Tensor
+
+
+__all__ = ['HybridizationPC']
+
+
+class HybridizationPC(PCBase):
+    """A Slate-based python preconditioner that solves a
+    mixed saddle-point problem using hybridization.
+
+    The forward eliminations and backwards reconstructions
+    are performed element-local using the Slate language.
+    """
+    def initialize(self, pc):
+        """Set up the problem context. Take the original
+        mixed problem and reformulate the problem as a
+        hybridized mixed system.
+
+        A KSP is created for the Lagrange multiplier system.
+        """
+        from ufl.algorithms.map_integrands import map_integrand_dags
+        from firedrake import (FunctionSpace, TrialFunction,
+                               TrialFunctions, TestFunction, Function,
+                               BrokenElement, MixedElement,
+                               FacetNormal, Constant, DirichletBC,
+                               Projector)
+        from firedrake.assemble import (allocate_matrix,
+                                        create_assembly_callable)
+        from firedrake.formmanipulation import ArgumentReplacer, split_form
```

Comment:
Some of the tools I need (`DirichletBC`, `Projector`/`project`, and everything from `assemble.py` to name a few) cause `ImportError`s if I try to import them from the beginning of the file. So I opted to import most of what I need here.

I don't think I'm doing anything too radical here. I'm importing a lot so I understand the possible eye-sore. If you think it's a big problem, I can move the imports which don't cause trouble.